src/cpp/configoptions.cpp

```diff
@@ -0,0 +1,216 @@
+void ConfigOptions::set_default_from_string(const std::string& json_blob)
+{
+    _default_source = cs_blob;
+    _default_cfg_string = json_blob;
+}
+
+// Retrieve the value of a numeric configuration option
+int64_t ConfigOptions::get_integer_option(
+    const std::string& key, int64_t default_value)
+{
+    // If we already have the value, return it
+    auto search = _int_options.find(key);
+    if (search != _int_options.end())
+        return _int_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    int64_t result = default_value;
+    if (_lazy) {
+        int temp = 0;
+        get_config_integer(temp, key, default_value);
+        result = (int64_t)temp;
+    }
+
+    // Store the final value before we exit
+    _int_options.insert({key, result});
+    return result;
+}
+
+// Retrieve the value of a string configuration option
+std::string ConfigOptions::get_string_option(
+    const std::string& key, const std::string& default_value)
+{
+    // If we already have the value, return it
+    auto search = _string_options.find(key);
+    if (search != _string_options.end())
+        return _string_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    std::string result(default_value);
+    if (_lazy) {
+        get_config_string(result, key, default_value);
+    }
+
+    // Store the final value before we exit
+    _string_options.insert({key, result});
+    return result;
+}
+
+// Retrieve the value of a boolean configuration option
+bool ConfigOptions::get_boolean_option(
+    const std::string& key, bool default_value)
+{
+    // If we already have the value, return it
+    auto search = _bool_options.find(key);
+    if (search != _bool_options.end())
+        return _bool_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    bool result = default_value;
+    if (_lazy) {
+        get_config_bool(result, key, default_value);
+    }
+
+    // Store the final value before we exit
+    _bool_options.insert({key, result});
+    return result;
+}
+
+// Check whether a configuration option is set in the selected source
+bool ConfigOptions::is_defined(const std::string& key)
+{
+    // Check each map in turn
+    if (_int_options.find(key) != _int_options.end())
+        return true;
+    if (_string_options.find(key) != _string_options.end())
+        return true;
+    if (_bool_options.find(key) != _bool_options.end())
+        return true;
+
+    // Not found
+    return false;
+}
+
+// Override the value of a numeric configuration option
+void ConfigOptions::override_integer_option(
+    const std::string& key, int64_t value)
+{
+    _int_options.insert_or_assign(key, value);
+}
+
+// Override the value of a string configuration option
+void ConfigOptions::override_string_option(
+    const std::string& key, const std::string& value)
+{
+    _string_options.insert_or_assign(key, value);
```

Comment:
it seems unnecessary to have these override methods if the underlying action is insert_or_assign. since they're already forced to make multiple steps to override, they might as well just do `cfg = build_cfg(); cfg.set('foo', value)` and you keep fewer methods on your API

alternatively, scrap this pattern and establish your config source hierarchy. you'd do it like... load ~/.smartsim/config.json, then override from env, then override from constructor. this way, clients can patch the defaults and write the least amount of code

Reply:
Per Andrew, this pattern has been enormously popular and successful with the MOM6 community, one of our main supporters. It allows baseline configurations to be centrally established, then shipped out to multiple users and/or experiments with local tweaking to finalize options.

Reply:
I was thinking that it could also force clients to think of config as idempotent... Why should we let them set anything at all? If we force config values to be frozen, they can be assured that the environment won't get changed out from under them by some rogue process

Reply:
@ankona: API simplicity is an entirely valid criticism of this approach, however in my opinion there is merit to having a 'one method' above all others to set these configuration options. While you are correct that the infrastructure might automatically set all of these things for the user beforehand, ultimately the user-code should be able to take precedence over all others (though they bear responsibility then for any unforeseen disallowed combinations/values).

Take for example the logging level where we know a priori that large numbers of clients this will be a performance hit due to I/O. Giving the users an 'override' method in application code to ensure that the logging level is 'quiet', no matter what else is going on is a simple guardrail.

Following this example, `override` is a little bit is thus shorthand for the fact that setting a config option via user-code is the highest priority. If we wanted to push `override` beyond simply being syntactic sugar, we could conceivably make this a 'set and lock' type operation, i.e. iff a config value is set by the `override` method, then trying to set the value again (by any means) is immediately a runtime option.